### PR TITLE
711 apply access checks for results

### PR DESCRIPTION
**Pull request by Dhruv-Copods**
Added backend checks for result features

### CODE REVIEW

File: app/routes/$workspaceId/results/groupByTests/$testId.$candidateId.$sectionId.tsx [hunk 1]
```diff
@@ -14,8 +17,20 @@ export type LoaderData = {
...
+  try {
+    const sectionDetail = await getResultDetailBySection(
+      params?.sectionId!,
+      userId!,
+      currentWorkspaceId
+    )
+    return json<LoaderData>({ params, sectionDetail })
+  } catch (error: any) {
+    if (error.status === HTTP_CODE.ACCESS_DENIED) {
+      return redirect(routes.members)
```
Comment (richa-copods) [inline app/routes/$workspaceId/results/groupByTests/$testId.$candidateId.$sectionId.tsx:31]: Create a simple Access Denied Page and redirect user to that Page, for all the redirects

File: app/models/result.server.ts [hunk 2]
```diff
@@ -1,406 +1,462 @@
...
-      candidateTest: {
-        some: {
-          id: {
-            not: undefined,
+  try {
+    if (
+      !(await checkFeatureAuthorization(userId, workspaceId, "results", "read"))
+    ) {
+      throw {
+        status: 403,
```
Comment (richa-copods) [inline app/models/result.server.ts:325]: Just status is sufficient, no need to send message and data field. For all the code blocks

File: app/interface/Interface.tsx [hunk 3]
```diff
@@ -419,6 +419,15 @@ export interface headerProps {
   rightChildren?: React.ReactNode
 }
 
+export enum HTTP_CODE {
```
Comment (richa-copods) [inline app/interface/Interface.tsx:422]: no need of this

**Review by richa-copods — CHANGES_REQUESTED** [review]

File: app/routes/$workspaceId/results/groupByTests/$testId.tsx [hunk 4]
```diff
@@ -19,34 +21,41 @@ export const loader: LoaderFunction = async ({ request, params }) => {
...
-    candidatesOfTest,
-    params,
-    workspaces,
-    currentWorkspaceId,
-    candidatesCount,
-    currentPage,
-    pageSize,
-  })
+  try {
+    const candidatesCount = await getALLCandidatesOfTestCount(
```
Comment (richa-copods) [inline app/routes/$workspaceId/results/groupByTests/$testId.tsx:30]: Shouldn't it be `getCountofAllCandidatesOfTest` or at least `getAllCandidatesOfTestCount`

File: app/routes/$workspaceId/results/groupByTests/index.tsx [hunk 5]
```diff
@@ -32,31 +32,43 @@ export const loader: LoaderFunction = async ({ request, params }) => {
...
-    userId,
-    workspaces,
-    currentWorkspaceId,
-    resultsCurrentPage,
-    resultsItemsPerPage,
-    testCount,
-    totalTestCount,
-  })
+  try {
+    const totalTestCount = await getTotalTestCounts(
```
Comment (richa-copods) [inline app/routes/$workspaceId/results/groupByTests/index.tsx:38]: Can be named `getTotalTestsCount`

File: public/locales/en/index.json [hunk 6]
```diff
@@ -358,5 +358,9 @@
     "notFound": "Not Found",
     "urlNotFound": "The requested URL was not found on this server.",
     "gotoDashboard": "GoTo Dashboard"
+  },
+  "403": {
```
Comment (richa-copods) [inline public/locales/en/index.json:362]: Name it to unauthorised

File: app/routes/index.tsx [hunk 7]
```diff
@@ -7,7 +7,7 @@ import Logo from "~/components/Logo"
 import { routes } from "~/constants/route.constants"
 import { getUserId } from "~/session.server"
 
-export const loader: LoaderFunction = async ({ request }) => {
+export const loader: LoaderFunction = async ({ request, params }) => {
```
Comment (richa-copods) [inline app/routes/index.tsx:10]: Why have you destructured `params` when not being used?

**Review by richa-copods — CHANGES_REQUESTED** [review]

File: app/routes/$workspaceId/results/groupByTests/index.tsx [hunk 8]
```diff
@@ -32,31 +32,43 @@ export const loader: LoaderFunction = async ({ request, params }) => {
...
+      params.testId!,
+      currentWorkspaceId,
+      userId
+    )
+    const testCount = await getALLCandidateTestsCount(
```
Comment (richa-copods) [inline app/routes/$workspaceId/results/groupByTests/index.tsx:43]: Shouldn't it be getCountofAllCandidatesOfTest or at least getAllCandidatesOfTestCount

File: app/routes/$workspaceId/results/groupByTests/index.tsx [hunk 9]
```diff
@@ -32,31 +32,43 @@ export const loader: LoaderFunction = async ({ request, params }) => {
...
+      currentWorkspaceId,
+      statusFilter,
+      userId
+    )
+    const candidateTest = await getALLCandidateTests(
```
Comment (richa-copods) [inline app/routes/$workspaceId/results/groupByTests/index.tsx:48]: Shouldn't it be `getAllCandidatesOfTest`

**Review by richa-copods — CHANGES_REQUESTED** [review]

File: app/routes/index.tsx (hunk 7 above)
Comment (Dhruv-Copods) [reply app/routes/index.tsx:10]: I had imported it while testing period, removing it

File: public/locales/en/index.json [hunk 10]
```diff
@@ -358,5 +358,9 @@
     "notFound": "Not Found",
     "urlNotFound": "The requested URL was not found on this server.",
     "gotoDashboard": "GoTo Dashboard"
+  },
+  "unauthorized": {
+    "error": "The route you tried to access was not authorized",
```
Comment (richa-copods) [inline public/locales/en/index.json:363]: "You are not authorised to access this route"

**Review by richa-copods — CHANGES_REQUESTED** [review]

File: public/locales/en/index.json [hunk 11]
```diff
@@ -358,5 +358,9 @@
     "notFound": "Not Found",
     "urlNotFound": "The requested URL was not found on this server.",
     "gotoDashboard": "GoTo Dashboard"
+  },
+  "unauthorized": {
+    "error": "You are not authorised to access this route",
+    "gotoDashboard": "GoTo Dashboard"
```
Comment (richa-copods) [inline public/locales/en/index.json:364]: "Go to Dashboard"

File: app/services/results.service.ts [hunk 12]
```diff
@@ -56,23 +63,39 @@ export const getWorkspaces = async (userId: string) => {
...
 }
 
 /**
  * this function will return all candidate tests count
  * @param currentWorkspaceId
  * @param statusFilter
+ * @param userId
  * @returns total count
  */
 export const getALLCandidateTestsCount = async (
```
Comment (richa-copods) [inline app/services/results.service.ts:89]: Rename this

File: app/services/results.service.ts [hunk 13]
```diff
@@ -18,6 +18,7 @@ import { getUserWorkspaces } from "~/models/workspace.server"
  * @param statusFilter
  * @param sortBy
  * @param sortOrder
+ * @param userId
  * @returns Test array
  */
 export const getALLCandidateTests = async (
```
Comment (richa-copods) [inline app/services/results.service.ts:24]: Rename this

File: public/locales/en/index.json (hunk 11 above)
Comment (richa-copods) [reply public/locales/en/index.json:364]: correct value of error for "404" also

**Review by richa-copods — CHANGES_REQUESTED** [review]

**Review by richa-copods — APPROVED** [review]

**Pull request merged**